[api] Add new class QgsGoogleMapsGeocoder

nyalldawson opened the pull request:
This geocoder utilises the Google Maps "geocoding" API in order to geocode
strings. The Google Maps service is not publicly available, and accordingly
an API key must be first obtained from Google and specified when constructing
this class.

(The user is responsible for managing their Google Maps API key, and ensuring
that the use of this geocoder does not exceed their usage limits! Excessive use
of the Google Maps geocoder API can result in charges being applied to the API key
holder.)

This is raw underlying API only. It is intended to be exposed to QGIS users only
via plugins, and does not result in any out-of-the-box Google Maps functionality.

In order for a plugin to use it, they must do something like this:

    # create a google maps geocoder
    api_key = 'my api key'
    coder = QgsGoogleMapsGeocoder(api_key)

    # add it to the locator bar
    filter = QgsGeocoderLocatorFilter('Google', 'Google', 'addr', coder, iface.mapCanvas())
    iface.registerLocatorFilter(filter)

Mini FAQ:

Q: Why is this being added to core, instead of a plugin?
A: While a plugin would be perfectly acceptable if we are only targetting QGIS desktop,
the intention here is to allow this underlying, low level class to be reused outside
of QGIS desktop (e.g. on QField) with a minimal amount of duplicate effort.

nyalldawson committed:
[api] Add new class QgsGoogleMapsGeocoder

This geocoder utilises the Google Maps "geocoding" API in order to geocode
strings. The Google Maps service is not publicly available, and accordingly
an API key must be first obtained from Google and specified when constructing
this class.

(The user is responsible for managing their Google Maps API key, and ensuring
that the use of this geocoder does not exceed their usage limits! Excessive use
of the Google Maps geocoder API can result in charges being applied to the API key
holder.)

This is raw underlying API only. It is intended to be exposed to QGIS users only
via plugins, and does not result in any out-of-the-box Google Maps functionality.

In order for a plugin to use it, they must do something like this:

    # create a google maps geocoder
    api_key = 'my api key'
    coder = QgsGoogleMapsGeocoder(api_key)

    # add it to the locator bar
    filter = QgsGeocoderLocatorFilter('Google', 'Google', 'addr', coder, iface.mapCanvas())
    iface.registerLocatorFilter(filter)

Mini FAQ:

Q: Why is this being added to core, instead of a plugin?
A: While a plugin would be perfectly acceptable if we are only targetting QGIS desktop,
the intention here is to allow this underlying, low level class to be reused outside
of QGIS desktop (e.g. on QField) with a minimal amount of duplicate effort.

nyalldawson commented:
Ping @3nids @rduivenvoorde @nirvn

3nids commented on src/core/geocoding/qgsgooglemapsgeocoder.h:
```diff
@@ -0,0 +1,83 @@
+
+  private:
+
+    QString mApiKey;
```
can we have setter/getter for API key and region?

nyalldawson replied:
Done

3nids commented on src/core/geocoding/qgsgooglemapsgeocoder.cpp:
```diff
@@ -0,0 +1,253 @@
+
+  const QgsGeometry geom = QgsGeometry::fromPointXY( QgsPointXY( longitude, latitude ) );
+
+  QgsGeocoderResult res( json.value( QStringLiteral( "formatted_address" ) ).toString(),
```
could we have a way to be a bit finer in the results and have a group and description properties as string? If given they would be used in the locator filter result.

nyalldawson replied:
I've added group, and the api for description. There's nothing returned by the google maps geocoder which lends itself to the description field though (rather that applies more to the Google places api, which could be a potential other geocoder class!)

3nids replied:
perfect, thanks a lot!

3nids reviewed (approved):
Nice!

nyalldawson committed:
Add optional viewport property for geocoder results

Allows specifying an optional recommended viewport bounds for
displaying the geocode result (e.g. the canvas extent to zoom
to for showing the results)

nyalldawson committed:
Add group and description to QgsGeocoderResult, link in to locator,
and populate group in Google Maps geocoding results

nyalldawson committed:
Add getter/setter for api key, region
